### PR TITLE
ubsan/memsan fix.

### PR DESCRIPTION
**Pull request by amosbird**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

- Bug Fix

When getting data from `ColumnArray`  https://github.com/yandex/ClickHouse/blob/master/dbms/src/Columns/ColumnArray.cpp#L144
we can get an invalid pointer when current array is empty. (due to left padding of podarray, it should be some global readonly address)

Theoretically we'd like to get the same pointer value in https://github.com/yandex/ClickHouse/blob/master/dbms/src/Columns/ColumnArray.cpp#L151 . However, since we have a dedicated ZeroValueStorage in hash table. The pointer of zero lengthed arrays becomes nil after placing into the agg states hash table, which leads to UBSan and MemSan failure.

### CODE REVIEW

**Comment by amosbird** [comment]
This bug can be reproduced using the following query

```
SELECT groupUniqArray(v)
FROM
(
    SELECT
        1 AS id,
        [2] AS v
    UNION ALL
    SELECT
        1 AS id,
        [] AS v
)
GROUP BY id
```

clickhouse build command is
```
cmake -DSANITIZE=memory -DCMAKE_C_COMPILER="clang-8" -DCMAKE_CXX_COMPILER="clang++-8" -DCMAKE_CXX_FLAGS="-stdlib=libc++" -DCMAKE_BUILD_TYPE=RelWithDebInfo -DCMAKE_CXX_FLAGS="-O1 -fno-omit-frame-pointer -fno-optimize-sibling-calls" -DCMAKE_C_FLAGS="-O1 -fno-omit-frame-pointer -fno-optimize-sibling-calls" -D ENABLE_HDFS=0 -D ENABLE_CAPNP=0 -D ENABLE_RDKAFKA=0 -D ENABLE_ICU=0 -D ENABLE_POCO_MONGODB=0 -D ENABLE_POCO_NETSSL=0 -D ENABLE_POCO_ODBC=0 -D ENABLE_ODBC=0 -D ENABLE_MYSQL=0 -D ENABLE_EMBEDDED_COMPILER=0 -D USE_INTERNAL_CAPNP_LIBRARY=0 -D USE_SIMDJSON=0 -DENABLE_READLINE=0 -DENABLE_ORC=0 -DENABLE_PARQUET=0 -DUSE_CAPNP=0 -DUSE_INTERNAL_ORC_LIBRARY=0 -DUSE_ORC=0 -DENABLE_SSL=0 ..
``

**Comment by amosbird** [comment]
Actuall we can also make all StringRef zero value storages pointing to a global empty string.

**Comment by akuzm** [comment]
Great, thanks for investigating!

Your test case reproduces the error for me, we probably should add id to the tests. It has a shorter form using the new Values table function: `select groupUniqArray(v) from values('id int, v Array(int)', (1, [2]), (1, [])) group by id`.

So, is the core cause that we compare uninitialized `end` to uninitialized `pos` here?
https://github.com/yandex/ClickHouse/blob/master/dbms/src/Columns/ColumnArray.cpp#L166

I think it would be good to achieve that the `pos` is always well-defined, by pointing it to a global empty string as you suggest. Or maybe even making it `nullptr` in such cases? That is, require that if `StringRef.size == 0`, then `StringRef.data == nullptr`.

**Comment by amosbird** [comment]
I guess it's because ZeroValueStorage is not initialized. https://github.com/yandex/ClickHouse/blob/master/dbms/src/Common/HashTable/HashTable.h#L223 
. I'm not sure how memsan reasons about this failure, but I remember the ubsan error is due to the invalid pos = 0xffffffffffffffff  and pos + field_size overflows.

**Comment by akuzm** [comment]
Looking at emplaceIfZero, I don't see where the constructor is called for zero value Cell at all. Is this correct?

Before, there was an explicit initialization of zero StringRef in GroupUniqArray:
https://github.com/yandex/ClickHouse/pull/6729/files#diff-9281c6b3757a84abfb5643adf2bb48d2L251

but I removed it in the key holder PR -- don't remember why.

Anyway, it looks like it was a hack, and we should just properly call the constructor for zero Cell, the same way we do for all other Cells in emplaceNonZeroImpl.

**Comment by amosbird** [comment]
yeah, initializing it properly also fixes the bug.

**Comment by akuzm** [comment]
I'd prefer to call the Cell constructor in emplaceIfZero, because it would be only initialized on demand, and also would properly initialize complex cells like ClearableHashTableCell, which is not initialized to all zeros but has a version field.

**Comment by amosbird** [comment]
@akuzm you just caught me in the middle :) . That commit was supposed to run a CI test but I soon realize it's too slow so I did locally...

**Comment by amosbird** [comment]
I cannot reproduce the thread san build failure locally using 

```
docker run -it --rm --net=host -v /ci/tsandebs:/package_folder  yandex/clickhouse-stateless-test
```

Is it some flaky test?

**Comment by akuzm** [comment]
> Is it some flaky test?

The logs are truncated, looks like some CI failure.

File: dbms/src/Common/HashTable/HashTable.h [hunk 1]
```diff
@@ -224,7 +224,11 @@ struct ZeroValueStorage<true, Cell>
 
 public:
     bool hasZero() const { return has_zero; }
-    void setHasZero() { has_zero = true; }
+    void setHasZero()
+    {
+        has_zero = true;
+        zero_value_storage = {};
```
Comment (akuzm) [inline dbms/src/Common/HashTable/HashTable.h]: I think we should call the Cell constructor here, like we do for non-zero Cells. We already have Cells with non-trivial constructors, like ClearableHashTableCell.

**Comment by amosbird** [comment]
Ok, how can I trigger a retest without force push new stuff?

File: dbms/src/Common/HashTable/HashTable.h (hunk 1 above)
Comment (amosbird) [reply dbms/src/Common/HashTable/HashTable.h]: ClearableHashTableCell  doesn't use zero storage

https://github.com/yandex/ClickHouse/blob/master/dbms/src/Common/HashTable/ClearableHashSet.h#L45

Comment (amosbird) [reply dbms/src/Common/HashTable/HashTable.h]: And I don't think constructing an std::aligned_storage_t as Cell even works...

**Comment by akuzm** [comment]
The thread sanitizer CI failure looks unrelated. The performance test didn't finish, but it shouldn't be influenced because we didn't change much. Will merge as is.

**Pull request merged**